src/python/pants/python/python_setup.py

```diff
@@ -216,6 +224,134 @@ def get_pex_python_paths():
         else:
             return []
 
+    @staticmethod
+    def get_asdf_paths(env: Environment, *, asdf_local: bool = False) -> List[str]:
+        """Returns a list of paths to Python interpreters managed by ASDF.
+
+        :param env: The environment to use to look up ASDF.
+        :param bool asdf_local: If True, only use the interpreter specified by
+                                '.tool-versions' file under `build_root`.
+        """
```

Comment:
Although our `pyenv` integration is not labeled with a TODO for it, both that code and this code should be using an "absolute filesystem local" API which doesn't currently exist: #10842. Without doing so, they are likely to have invalidation issues, where `pantsd` does not notice that the `pyenv`/`asdf` version has changed.

To get around this, `@rules` which touch `PythonSetup.interpreter_search_paths` should probably be marked with `@_uncacheable_rule` rather than `@rule`: that primarily appears to be [find_pex_python](https://github.com/pantsbuild/pants/blob/b0b7b0b914fd9df86be64e0948a87a01ccbce528/src/python/pants/backend/python/util_rules/pex_environment.py#L127-L133).

Since this was an existing issue, we'd be happy to fix it after this change lands.

Reply:
I appreciate the help. If there is something I can do on the ASDF side of things to assist when #10842 is resolved, I'd be happy to help.